### ISSUE TITLE
Unified Angular Location Service incorrectly parses the initial URL when in useHash mode

### ISSUE THREAD

**Issue by christopherthielen** [angular/angular#34600]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/common/upgrade

### Is this a regression?
probably not

### Description

When:
- The Unified Location Service is being used
- The `{ useHash: true }` mode is active
- During bootstrap, the browser URL contains an initial route in the hash, such as `http://mydomain.com/#/initialRoute`

Then:
I expect the $location path to be `/initialRoute`.
However, the $location `path` is `/` and the `hash` is `/initialRoute`

## 🔬 Minimal Reproduction
[This stackblitz](https://stackblitz.com/edit/angular-hybrid-uniified-location-initial-path?file=angularjs.module.ts) demonstrates the problem.  
When it loads, click the button labelled "Reload page with an initial path in the hash".

The stackblitz also has a toggle between the `$locationShim` from `@angular/common/upgrade` and the old AngularJS `$location` service showing the difference in behavior parsing the initial URL.

## 🌍  Your Environment

**Angular Version: 8**
with @angular/upgrade

**Anything else relevant?**

This issue is related to #34596.

**Comment by gkalpak** [comment]
This is still an issue in v9. It might have the same root cause as #31882.

**Comment by pravinkumart** [comment]
I have the same issue and I think in the following method we need to strip hash before calling parseAppURL method as like angularjs does
`$$parse(url) {
        /** @type {?} */
        let pathUrl;
        if (url.startsWith('/')) {
            pathUrl = url;
        }
        else {
            // Remove protocol & hostname if URL starts with it
            pathUrl = this.stripBaseUrl(this.getServerBase(), url);
        }
        if (typeof pathUrl === 'undefined') {
            throw new Error(`Invalid url "${url}", missing path prefix "${this.getServerBase()}".`);
        }
        this.parseAppUrl(pathUrl);
        if (!this.$$path) {
            this.$$path = '/';
        }
        this.composeUrls();
    }`

**Comment by mfp22** [comment]
@pravinkumart You are right, but there is an extra thing that is needed in some cases. In my app, it isn't just a `/` before the `#`, but a path like `localhost:9797/dev/#/route`. The whole `/dev/` part needs to be stripped, so when I put this line:
```
pathUrl = pathUrl.slice(pathUrl.indexOf('#') + 1);
```
above `this.parseAppUrl(pathUrl);` the app loads correctly initially and other navigations
